DOC-5801: search: add new FT.HYBRID command (#2210)
redis/docs: content/commands/ft.hybrid.md

configures K-nearest neighbors search for vector similarity. The `count` parameter indicates the number of following parameters. `K` specifies the number of nearest neighbors to find. `EF_RUNTIME` controls the search accuracy vs. speed tradeoff. `YIELD_DISTANCE_AS` assigns an alias to the distance value.
</details>

<details open>
<summary><code>RANGE count RADIUS radius-value [EPSILON epsilon-value] [YIELD_DISTANCE_AS distance_field]</code></summary>

configures range-based vector search within a specified radius. The `count` parameter indicates the number of following parameters. `RADIUS` defines the maximum distance for matches. `EPSILON` provides additional precision control.
</details>

<details open>
<summary><code>FILTER "filter-expression"</code></summary>

applies pre-filtering to vector search results. This filter affects which documents are considered for vector similarity but doesn't impact scoring, unlike the `SEARCH` component which affects both filtering and scoring.
</details>

<details open>
<summary><code>POLICY [ADHOC|BATCHES|ACORN] [BATCH_SIZE batch-size-value]</code></summary>

> controls the pre-filtering policy for vector queries. `ADHOC` processes filters on-demand, `BATCHES` processes in configurable batch sizes, and `ACORN` provides future support for advanced pre-filtering.

Reviewer: ```suggestion
controls the pre-filtering policy for vector queries. `ADHOC_BF` processes filters on-demand and `BATCHES` processes in configurable batch sizes, refer to the pre-filtering policy [here](https://redis.io/docs/latest/develop/ai/search-and-query/vectors/#filters).
```

## Default values and behaviors

FT.HYBRID provides sensible defaults to ease onboarding:

- **Query count**: 2 (one SEARCH and one VSIM component required)
- **Default LIMIT**: 10 results
- **Default SCORER**: BM25STD for text search
- **Default KNN K**: 10 neighbors
- **Default RRF WINDOW**: 20 (or follows LIMIT if specified)
- **Default RRF CONSTANT**: 60 (following Elasticsearch convention)

## Parameter count convention

All multi-parameter options use a count prefix that contains ALL tokens that follow:

- `SCORER 4 BM25 1.2 0.75` - algorithm + 2 parameter values
- `KNN 4 K 10 EF_RUNTIME 100` - 2 key-value pairs
- `PARAMS 4 min_price 50 max_price 200` - 2 key-value pairs
- `COMBINE RRF 4 WINDOW 40 CONSTANT 1.5` - RRF method with 2 key-value pairs

The only exception is alias usage with `AS`, which is not counted:
- `APPLY "@vector_distance+@score" AS final_score`
- `LOAD 3 category AS cat brand AS brd price AS prc`

## Reserved fields

The following fields are reserved for internal use:

- `@__key` - reserved for loading key IDs when required
- `@__score` - reserved for the combined score (can be aliased)
- `@vector_distance` - yields the vector distance (can be aliased)

> - `@__combined_score` - fused score from the COMBINE step

Reviewer: ```suggestion
```

  YIELD_DISTANCE_AS vector_dist
  COMBINE RRF 4 WINDOW 50 CONSTANT 80
  YIELD_SCORE_AS hybrid_score
  SORTBY hybrid_score DESC
  LIMIT 0 20
{{< / highlight >}}
</details>

<details open>
<summary><b>Hybrid search with pre-filtering</b></summary>

Search with vector pre-filtering and post-processing:

{{< highlight bash >}}
127.0.0.1:6379> FT.HYBRID products-idx
  SEARCH "smartphone"
  VSIM @image_vector $query_vec
  KNN 2 K 15
  FILTER "@price:[100 500]"
  COMBINE LINEAR 4 ALPHA 0.7 BETA 0.3
  LOAD 4 title price category rating
  APPLY "@price * 0.9" AS discounted_price
  SORTBY rating DESC
{{< / highlight >}}
</details>

<details open>
<summary><b>Hybrid search with parameters</b></summary>

Use parameter substitution for dynamic queries:

{{< highlight bash >}}
127.0.0.1:6379> FT.HYBRID products-idx
  SEARCH "@brand:$brand_name"
  VSIM @content_vector $query_vector
  RANGE 4 RADIUS 0.8 EPSILON 0.1
  FILTER "@availability:$stock_status"
  PARAMS 6 brand_name "Apple" query_vector <vector_blob> stock_status "in_stock"
  EXPLAINSCORE
{{< / highlight >}}
</details>

## Complexity

FT.HYBRID complexity depends on both the text search and vector similarity components:

> - Text search: O(n) where n is the number of matching documents

Reviewer: ```suggestion
- Text search: O(n) for simple term searches, where n is the number of matching documents. In multi-term queries with INTERSECT or UNION, or when using fuzzy or prefix matches, complexity increases proportionally to the total number of entries scanned across all participating terms.
```

| Redis<br />Enterprise | Redis<br />Cloud | <span style="min-width: 9em; display: table-cell">Notes</span> |
|:----------------------|:-----------------|:------|
| <span title="Not supported">&#x274c; Standard</span><br /><span title="Not supported"><nobr>&#x274c; Active-Active</nobr></span> | <span title="Not supported">&#x274c; Standard</span><br /><span title="Not supported"><nobr>&#x274c; Active-Active</nobr></span> |  |

## Return information

{{< multitabs id="ft-hybrid-return-info" 
    tab1="RESP2" 
    tab2="RESP3" >}}

One of the following:
* [Array]({{< relref "/develop/reference/protocol-spec#arrays" >}}) with the first element being the total number of results, followed by document IDs and their field-value pairs as [arrays]({{< relref "/develop/reference/protocol-spec#arrays" >}}).
* [Simple error reply]({{< relref "/develop/reference/protocol-spec#simple-errors" >}}) in these cases: no such index, syntax error in query.

-tab-sep-

One of the following:
* [Map]({{< relref "/develop/reference/protocol-spec#maps" >}}) with the following fields:
    - `total_results`: [Integer]({{< relref "/develop/reference/protocol-spec#integers" >}}) - total number of results

>     - `results`: [Array]({{< relref "/develop/reference/protocol-spec#arrays" >}}) of [maps]({{< relref "/develop/reference/protocol-spec#maps" >}}) containing document information

Reviewer: ```suggestion
    - `execution_time`: [doubles]({{< relref "/develop/reference/protocol-spec#doubles" >}}) containing hybrid query execution time
```
